flink-core/src/main/java/org/apache/flink/api/dag/Transformation.java

```diff
@@ -139,12 +142,12 @@ public static int getNewNodeId() {
 	private ResourceSpec preferredResources = ResourceSpec.DEFAULT;
 
 	/**
-	 * This weight indicates how much this transformation relies on managed memory, so that
-	 * transformation highly relies on managed memory would be able to acquire more managed
-	 * memory in runtime (linear association). Note that it only works in cases of UNKNOWN
-	 * resources.
+	 * Each entry in this map represents a use case that this transformation needs managed memory for. The key of the
+	 * entry indicates the use case, while the value is the use-case-specific weight for this transformation. Managed
+	 * memory reserved for an OPERATOR scope use case will be shared by all the declaring transformations within a slot
+	 * according to this weight. For SLOT scope use cases, only the keys matter and the values (weights) are ignored.
 	 */
-	private int managedMemoryWeight = DEFAULT_MANAGED_MEMORY_WEIGHT;
+	private final Map<ManagedMemoryUseCase, Integer> managedMemoryUseCaseWeights = new HashMap<>();
```

Comment:
I would consider splitting
`managedMemoryUseCaseWeights` and
`declareManagedMemoryUseCase`, `getManagedMemoryUseCaseWeights`
into:
`managedMemoryOperatorUseCaseWeights/managedMemorySlotUseCaseWeights` and
`declareManagedMemoryOperatorUseCase(uc, w)/declareManagedMemorySlotUseCase(uc)`,
`getManagedMemoryOperatorUseCaseWeights/getManagedMemorySlotUseCases`
also for `StreamNode` in the next commit

Reply:
I think you are right.
The semantic that weights are ignored for slot scope use cases is a bit fragile and could easily be overlooked. Separating them into different maps/interfaces would be more explicit.